server/services/draftsService.test.ts

```diff
@@ -0,0 +1,174 @@
+import { Request } from 'express'
+import { mocked } from 'ts-jest/utils'
+import uuid from 'uuid'
+import DraftsService from './draftsService'
+
+jest.mock('uuid')
+const mockedUuid = mocked(uuid)
+
+describe(DraftsService, () => {
+  function createRequestWithSession(session: unknown): Request {
+    return { session } as unknown as Request
+  }
+
+  const clock = { now: () => new Date('2021-04-01T10:25:00Z') }
+
+  describe('.createDraft', () => {
+    it('generates a random UUID and stores the data in the session under a key derived from that UUID', () => {
+      mockedUuid.v4.mockReturnValue('2dc7ef56-58dd-4339-9924-c33318738068')
+      const req = createRequestWithSession({})
+
+      const draftsService = new DraftsService(clock)
+
+      const draft = draftsService.createDraft('someExampleType', { a: 1, b: 2 }, req)
+
+      expect(draft).toEqual({
+        id: '2dc7ef56-58dd-4339-9924-c33318738068',
+        type: 'someExampleType',
+        createdAt: new Date('2021-04-01T10:25:00Z'),
+        updatedAt: new Date('2021-04-01T10:25:00Z'),
+        data: { a: 1, b: 2 },
+      })
+
+      expect(req.session['draft-2dc7ef56-58dd-4339-9924-c33318738068']).toEqual({
+        id: '2dc7ef56-58dd-4339-9924-c33318738068',
+        type: 'someExampleType',
+        createdAt: '2021-04-01T10:25:00.000Z',
+        updatedAt: '2021-04-01T10:25:00.000Z',
+        data: { a: 1, b: 2 },
+      })
+    })
+  })
+
+  describe('.fetchDraft', () => {
+    describe('when the session contains a draft for that ID', () => {
+      it('returns that draft', () => {
+        const req = createRequestWithSession({
+          'draft-2dc7ef56-58dd-4339-9924-c33318738068': {
+            id: '2dc7ef56-58dd-4339-9924-c33318738068',
+            type: 'someExampleType',
+            createdAt: '2021-04-01T10:25:00Z',
+            updatedAt: '2021-04-01T10:25:00Z',
+            data: { a: 1, b: 2 },
+          },
+        })
+
+        const draftsService = new DraftsService(clock)
+
+        const draft = draftsService.fetchDraft('2dc7ef56-58dd-4339-9924-c33318738068', req)
+
+        expect(draft).toEqual({
+          id: '2dc7ef56-58dd-4339-9924-c33318738068',
+          type: 'someExampleType',
+          createdAt: new Date('2021-04-01T10:25:00Z'),
+          updatedAt: new Date('2021-04-01T10:25:00Z'),
+          data: { a: 1, b: 2 },
+        })
+      })
+    })
+
+    describe('when the session doesn’t contain a draft for that ID', () => {
+      it('throws an error', () => {
+        const req = createRequestWithSession({
+          'draft-2dc7ef56-58dd-4339-9924-c33318738068': {
+            id: '2dc7ef56-58dd-4339-9924-c33318738068',
+            type: 'someExampleType',
+            createdAt: '2021-04-01T10:25:00Z',
+            updatedAt: '2021-04-01T10:25:00Z',
+            data: { a: 1, b: 2 },
+          },
+        })
+
+        const draftsService = new DraftsService(clock)
+
+        expect(() => draftsService.fetchDraft('def456', req)).toThrow()
```

Comment:
how often will calls want to do fetches to check existence? e.g. try { fetchDraft(...) } catch { // no draft, do something else }

if it's a pattern we expect, it's more efficient to return a value here e.g. `null` rather than throw an execption.

Reply:
I think that the "something else" will always be to show some sort of an error page. But, as I've mentioned in one of the commits, the error handling should probably be per-journey anyway, to show useful information about how to restart the journey. So probs makes sense to get rid of the generic error and handle on a case-by-case basis.